apps/vscode/command_client.py

```diff
@@ -1,126 +1,235 @@
-    command: str,
+    command_id: str,
     *args: str,
     wait_for_finish: bool = False,
-    expect_response: bool = False,
+    return_command_output: bool = False,
     decode_json_arguments: bool = False,
 ):
-    """Execute command via vscode command server."""
+    """Runs a VSCode command, using command server if available
+
+    Args:
+        command (str): The ID of the VSCode command to run
+        wait_for_finish (bool, optional): Whether to wait for the command to finish before returning. Defaults to False.
+        return_command_output (bool, optional): Whether to return the output of the command. Defaults to False.
+        decode_json_arguments (bool, optional): Whether to decode JSON arguments. Defaults to False.
+
+    Raises:
+        Exception: If there is an issue with the file-based communication
+
+    Returns:
+        Object: The response from the command, if requested.
+    """
     # NB: This is a hack to work around the fact that talon doesn't support
     # variable argument lists
-    args = list(
-        filter(
-            lambda x: x is not NotSet,
-            args,
-        )
-    )
+    args = [x for x in args if x is not NotSet]
+
     if decode_json_arguments:
         args = [json.loads(arg) for arg in args]
 
-    port_file_path = Path(gettempdir()) / "vscode-port"
+    username = getpass.getuser()
 
-    if not port_file_path.exists():
-        if args or expect_response or decode_json_arguments:
+    communication_dir_path = Path(gettempdir()) / f"vscode-command-server-{username}"
+
+    if not communication_dir_path.exists():
+        if args or return_command_output or decode_json_arguments:
             raise Exception("Must use command-server extension for advanced commands")
         print("Port file not found; using command palette")
-        run_vscode_command_by_command_palette(command)
+        run_vscode_command_by_command_palette(command_id)
         return
 
-    original_contents = port_file_path.read_text()
+    request_path = communication_dir_path / "request.json"
+    response_path = communication_dir_path / "response.json"
+
+    uuid = str(uuid4())
+
+    unlink_if_exists(response_path)
+
+    request = Request(
+        command_id=command_id,
+        args=args,
+        wait_for_finish=wait_for_finish,
+        return_command_output=return_command_output,
+        uuid=uuid,
+    )
+
+    write_request(
+        request,
+        request_path,
+    )
 
     # Issue command to VSCode telling it to update the port file.  Because only
     # the active VSCode instance will accept keypresses, we can be sure that
     # the active VSCode instance will be the one to write the port.
-    if is_mac:
-        actions.key("cmd-shift-alt-p")
-    else:
-        actions.key("ctrl-shift-alt-p")
+    actions.user.trigger_command_server_command_execution()
+
+    try:
+        decoded_contents = read_json_with_timeout(response_path)
+    finally:
+        unlink_if_exists(request_path)
```

Comment:
Instead of unlinking the request on the Python side, the vscode side can unlink the request once it's done reading it. That will make it more obvious which side failed if there are some stale files sitting around, and makes it less likely to accidentally double-execute a request.

Reply:
VSCode does unlink the request once it's done reading; this code is just supposed to make talon clean up after itself in case of a broken VSCode.  But I do see your point around double-executing a request.

How about if instead of silently unlinking, we throw an exception if the file exists?  Then I guess we could either unlink the file or not before throwing the exception.  I'd be inclined to unlink, because otherwise on next request we'll throw an exception saying that another talon instance is running because there's a non-stale request file.  But as you mention, that's slightly less defensive against double-executing a request.

So I'd say either way I'd push for throwing an exception if request file still exists, but could go either way wrt unlinking.

Reply:
There's no guarantee the new file is related to your request, and this state generally shouldn't happen. I really don't think Python should unlink the request file. If vscode stops unlinking the request file for some reason, you should fix vscode. Python unlinking the request file removes an indicator something is wrong with the system. It also violates the exclusive lock guarantees for detecting 2+ in-flight requests, because you can feasibly write w/ 'x', unlink, write w/ 'x' without vscode having processed anything yet.